### This is an update on the progress made so far in fixing the ACS accounts.

In [1]:
import pandas as pd
import matplotlib as plt
import numpy as np

In [2]:
pwd

'C:\\Users\\19713\\Desktop\\ACS\\ACS Account Fix'

In [3]:
df2 = pd.read_excel('C:\\Users\\19713\\Desktop\\ACS\\ACS Account Fix\\df2.xlsx', index_col = 0)

This is a dataset containing the most updated work so far.

In [4]:
df2

Patient ID               Patient Name  Referral Source ID  \
0     0018V00002dY52LQAS         Freddie Elliott JR                 NaN   
1     0018V00002dYB8EQAW                test tester                 NaN   
2     0018V00002deoaoQAA            Derrisha Sutton                 NaN   
3     0018V00002deoapQAA        (Yuri) Yuriel Leyva                 NaN   
4     0018V00002deoaqQAA              Abdullahi Ali                 NaN   
...                  ...                        ...                 ...   
3062  001RP000006JCx0YAG  Johana Mota Lopez (minor)                 NaN   
3063  001RP000006JCx1YAG   Eliza Mota Lopez (minor)                 NaN   
3064  001RP000006JCx2YAG               Rosa Sanchez                 NaN   
3065  001RP000006JCx3YAG    Nilofar Nikpour-Farrokh                 NaN   
3066  001RP000006JCx4YAG                   Jane Kim  a028V00000ZpXmFQAV   

         Original Referral Text Preferred Referral Source  \
0                           NaN                       NaN   
1                           NaN                       NaN   
2                           NaN                       NaN   
3                           NaN                       NaN   
4                           NaN                       NaN   
...                         ...                       ...   
3062  Yesenia C. Seguro Noreste           Seguro Noroeste   
3063  Yesenia C. Seguro Noreste           Seguro Noroeste   
3064           Previous Patient            Return Patient   
3065           Previous Patient            Return Patient   
3066                 David Eder      Thuemmel Uhle & Eder   

                                        Alix's comments  
0                                                   NaN  
1                                                   NaN  
2                                                   NaN  
3                                                   NaN  
4                                                   NaN  
...                                                 ...  
3062  *need to know specific office to link them.. M...  
3063  *need to know specific office to link them.. M...  
3064                                                NaN  
3065                                                NaN  
3066                                                NaN  

[3067 rows x 6 columns]

In [5]:
df2[df2['Original Referral Text'].isnull()]

Patient ID         Patient Name Referral Source ID  \
0    0018V00002dY52LQAS   Freddie Elliott JR                NaN   
1    0018V00002dYB8EQAW          test tester                NaN   
2    0018V00002deoaoQAA      Derrisha Sutton                NaN   
3    0018V00002deoapQAA  (Yuri) Yuriel Leyva                NaN   
4    0018V00002deoaqQAA        Abdullahi Ali                NaN   
..                  ...                  ...                ...   
178  001RP000006JCqLYAW         MOISES PEREZ                NaN   
179  001RP000006JCtKYAW      Lidia Putintsev                NaN   
180  001RP000006JCwNYAW         Sean Merrick                NaN   
181  001RP000006VMX6YAO           John Smith                NaN   
182  001RP000006J8GsYAK       Amaya Espinoza                NaN   

    Original Referral Text Preferred Referral Source Alix's comments  
0                      NaN                       NaN             NaN  
1                      NaN                       NaN             NaN  
2                      NaN                       NaN             NaN  
3                      NaN                       NaN             NaN  
4                      NaN                       NaN             NaN  
..                     ...                       ...             ...  
178                    NaN                       NaN             NaN  
179                    NaN                       NaN             NaN  
180                    NaN                       NaN             NaN  
181                    NaN                       NaN             NaN  
182                    NaN                       NaN             NaN  

[183 rows x 6 columns]

The first 183 rows in the dataset have a null value in the "Original Referral Text" column, which is the column in which the marketers were inputting their referral sources. The rest of the rows have entries in this column. The column 'Preferred Referral Source' contains values which represent what the original referral texts will actually be in Salesforce. For example, instead of writing 'Yesenia C. Seguro Noreste,' this referral source will show up as "Seguro Noroeste" in Salesforce.

In [6]:
df2[df2['Preferred Referral Source'].isnull() & ~(df2['Original Referral Text'].isnull())]

Patient ID       Patient Name Referral Source ID  \
201   001RP000006IvWEYA0    Thanh Trang Luu                NaN   
202   001RP000006IvWFYA0             Vy Luu                NaN   
203   001RP000006IvWGYA0         Trinh Tran                NaN   
293   001RP000006IwlgYAC       Kelsey Allen                NaN   
308   001RP000006IwlvYAC       Bianca Perez                NaN   
...                  ...                ...                ...   
2712  001RP000006JCpiYAG     ALONDRA GARCIA                NaN   
2765  001RP000006JCqbYAG  Ruth Frederickson                NaN   
2807  001RP000006JCrHYAW       JEFFREY WOOD                NaN   
2994  001RP000006JCvtYAG         JACEY LANE                NaN   
3018  001RP000006JCwHYAW      JOHN HUMPHREY                NaN   

        Original Referral Text Preferred Referral Source Alix's comments  
201                   Sonny Do                       NaN         Online?  
202                   Sonny Do                       NaN         Online?  
203                   Sonny Do                       NaN         Online?  
293                       Mike                       NaN         Online?  
308            Marketing Team?                       NaN         Online?  
...                        ...                       ...             ...  
2712            Luis Martinez:                       NaN         Online?  
2765                   Call In                       NaN         Online?  
2807                    Lauren                       NaN         Online?  
2994                    Lauren                       NaN         Online?  
3018  Attorney/ Lily Marketing                       NaN         Online?  

[62 rows x 6 columns]

In [7]:
(df2[df2['Preferred Referral Source'].isnull() & ~(df2['Original Referral Text'].isnull())]).shape[0]

62

There are currently 62 cells in the "Preferred Referral Source" column which do not contain a value that also has a cell value in the "Original Referral Source" column.

In [8]:
df2[df2['Preferred Referral Source'].isnull() & ~(df2['Original Referral Text'].isnull()) & (df2['Alix\'s comments'] == 'Online?')]['Original Referral Text'].unique()

array(['Sonny Do', 'Mike', 'Marketing Team?', 'Work Relation',
       "Dr. Matt'S Hop Festival Committee Board", 'Employee - Nectar',
       'Marketing Team: Sarah', 'Call In', 'Job', 'Staff',
       'Wrote Internet But May Have Been Referred By Priority Life Chiro',
       'Rose Urgent Care: Sarah', 'Accident Care Se Front Desk', 'Ad',
       'Chamber Of Commerce', 'Sara: Marketing Team', 'Referral',
       'Beaverton Office Gave Me Info', 'Attorney (Tony)', 'Work', 'Tony',
       'Wife/Google/State Farm', 'R&R Velma Dr. Matt Referral Group',
       'Referral By Tony', 'Attorney: David Wales', 'Lazer Augi (Boss)',
       'Attorney- Sonny', 'Lauren Neuman', 'Jason Hargrave',
       'Aleida Gonzalez / Autobody Shop (Villegas Autobody)',
       'Rodrigo Morales', 'Lorena Santana', 'Kelly Whitlok',
       'Luis Martinez:', 'Lauren', 'Attorney/ Lily Marketing'],
      dtype=object)

The above contains original referral texts for which Alix commented 'Online?'. We should verify exactly how we should move forward with these accounts.

In [9]:
file = 'ACS account fix.xlsx'
sheet2 = pd.read_excel(file, sheet_name = 1, index_col = 0)

In [10]:
sheet2.reset_index(inplace = True)

In [11]:
keys = sheet2['Name'].values
values = sheet2['Id'].values
dict1 = dict(zip(keys, values))

In [12]:
# This dictionary contains all of the current referral sources and their corresponding ids
dict1

{'State Farm Ted Ferry South': 'a028V00000UKFLWQA5',
 'SAFE Transportation': 'a028V00000UKJ3EQAX',
 'Christopher Pierce, DC': 'a028V00000UKJ3iQAH',
 'Carr Used & Carr Auto Body': 'a028V00000UKJ4WQAX',
 'CARR Chevrolet': 'a028V00000UKJ4vQAH',
 'Powell Motors Inc': 'a028V00000UKJ5eQAH',
 "Rob's Automotive Repair": 'a028V00000UKJ5fQAH',
 'Columbia View Family Health Center': 'a028V00000UKJ68QAH',
 'Classic Collision - NE 11th Portland': 'a028V00000UKJ6mQAH',
 'Classic Collision - SE Powell': 'a028V00000UKJ6nQAH',
 'Classic Collision - SE Foster': 'a028V00000UKJ71QAH',
 'Classic Collision - Macadam Ave': 'a028V00000UKJ76QAH',
 'Classic Collision - Hillsboro': 'a028V00000UKJ7BQAX',
 'Classic Collision - Newberg': 'a028V00000UKJ7CQAX',
 'Classic Collision - Wilsonville': 'a028V00000UKJ7QQAX',
 'Country Financial Chemawa': 'a028V00000UKK8ZQAX',
 'Farmers(Ryan Brown': 'a028V00000UKOpvQAH',
 'High street Insurance & Finance Service': 'a028V00000UKSKwQAP',
 'Brady Mertz Law Office': 'a028V00000U

This is a dictionary containing all of the current referral sources and their corresponding IDs. We can use this dictionary to map already existing referral sources in our accounts to their corresponding referral source IDs.

In [13]:
df2[~df2['Referral Source ID'].isnull()]

Patient ID           Patient Name  Referral Source ID  \
186   001RP000006IvVzYAK             Keri Armer  a028V00000Zy8LLQAZ   
188   001RP000006IvW1YAK      Freddie Elliottje  a028V00000Zy8LLQAZ   
190   001RP000006IvW3YAK  Daniela Tirado Osovio  a028V00000Zy8LLQAZ   
191   001RP000006IvW4YAK    Juan Lagunas Garcia  a028V00000ZpXd0QAF   
193   001RP000006IvW6YAK           Josh Kremers  a028V00000Zy8LLQAZ   
...                  ...                    ...                 ...   
3053  001RP000006JCwrYAG          Jose Villegas  a028V00000Zy8LLQAZ   
3055  001RP000006JCwtYAG          JOSE VILLEGAS  a028V00000Zy8LLQAZ   
3056  001RP000006JCwuYAG        YOLANDA RAMIREZ  a028V00000Zy8LLQAZ   
3059  001RP000006JCwxYAG           Dorka Tefera  a028V00000Zy8LLQAZ   
3066  001RP000006JCx4YAG               Jane Kim  a028V00000ZpXmFQAV   

                      Original Referral Text  \
186                                   Google   
188                                   Google   
190                                   Google   
191                           Atty: Mike Dye   
193                                   Google   
...                                      ...   
3053                                  Google   
3055                                  Google   
3056                                  Google   
3059  Google Reviews, Lives Down The Street.   
3066                              David Eder   

                Preferred Referral Source Alix's comments  
186                                Google             NaN  
188                                Google             NaN  
190                                Google             NaN  
191   Law Offices of Michael B. Dye-Salem             NaN  
193                                Google             NaN  
...                                   ...             ...  
3053                               Google             NaN  
3055                               Google             NaN  
3056                               Google             NaN  
3059                               Google             NaN  
3066                 Thuemmel Uhle & Eder             NaN  

[903 rows x 6 columns]

This is the dataset filtered only with accounts which have referral source IDs.

In [14]:
df2[~df2['Referral Source ID'].isnull()].shape[0]

903

In [15]:
df2.shape[0]

3067

In [16]:
df2.shape[0]-df2[~df2['Referral Source ID'].isnull()].shape[0]

2164

There are 903 accounts which have referral source IDs, which means that there are 2164 accounts which do not. Given that the original dataset only contained 3 accounts which had referral source IDs, this means that we have made good progress so far. How many distinct referral sources must we create IDs for?

In [17]:
df2.loc[(df2['Referral Source ID'].isnull()) & ~(df2['Preferred Referral Source'].isnull())]['Preferred Referral Source'].unique()

array(['Return Patient', 'Online', 'AC Marketing Team: Miguel',
       'Attorney', 'AC Marketing Team: Andrea', 'Patient Referral',
       'Google Appointment', 'Gatti Law Firm',
       'AC Marketing Team: Estefania', 'Alpine Law Group', 'Insurance',
       'The Advocates Injury Attorneys', 'Primary Care Provider',
       'ACS Employee Referral', 'Drive By/Walk In', 'Avrek Law Firm',
       'Transfer', 'AFC Urgent Care', 'LeDoux Insurance',
       'AC Marketing Team: Lily', 'The Joint Chiropractic', 'Yelp',
       'ACS Employee', 'AC Marketing Team: Yesenia',
       'AC Marketing Team: Karla', 'AC Marketing Team: Alma', 'Other',
       'Vietnamese Newspaper', 'AC Marketing Team: Nancy',
       'Seguro Noroeste', 'ZoomCare', 'Columbia Urgent Care',
       'Medical Office', 'AC Marketing Team: Alix', 'Youd Law',
       'AC Marketing Team: Dr. Alexis', 'Concentra',
       'AC Marketing Team: Dr. Linda', 'Auto Body',
       'BestMed (Clinic Unknown)', 'Bishop Legal', 'Bond Legal',
       '

In [18]:
len(df2.loc[(df2['Referral Source ID'].isnull()) & ~(df2['Preferred Referral Source'].isnull())]['Preferred Referral Source'].unique())

57

Two cells above we have an array of referral sources that do not yet have an associated ID. There are only 57 sources that we would have to create in order to provide IDs for those accounts which did have an original referral source text. The goal now is to develop a plan to account for these 57 sources which still lack ID numbers.

In [19]:
df2

Patient ID               Patient Name  Referral Source ID  \
0     0018V00002dY52LQAS         Freddie Elliott JR                 NaN   
1     0018V00002dYB8EQAW                test tester                 NaN   
2     0018V00002deoaoQAA            Derrisha Sutton                 NaN   
3     0018V00002deoapQAA        (Yuri) Yuriel Leyva                 NaN   
4     0018V00002deoaqQAA              Abdullahi Ali                 NaN   
...                  ...                        ...                 ...   
3062  001RP000006JCx0YAG  Johana Mota Lopez (minor)                 NaN   
3063  001RP000006JCx1YAG   Eliza Mota Lopez (minor)                 NaN   
3064  001RP000006JCx2YAG               Rosa Sanchez                 NaN   
3065  001RP000006JCx3YAG    Nilofar Nikpour-Farrokh                 NaN   
3066  001RP000006JCx4YAG                   Jane Kim  a028V00000ZpXmFQAV   

         Original Referral Text Preferred Referral Source  \
0                           NaN                       NaN   
1                           NaN                       NaN   
2                           NaN                       NaN   
3                           NaN                       NaN   
4                           NaN                       NaN   
...                         ...                       ...   
3062  Yesenia C. Seguro Noreste           Seguro Noroeste   
3063  Yesenia C. Seguro Noreste           Seguro Noroeste   
3064           Previous Patient            Return Patient   
3065           Previous Patient            Return Patient   
3066                 David Eder      Thuemmel Uhle & Eder   

                                        Alix's comments  
0                                                   NaN  
1                                                   NaN  
2                                                   NaN  
3                                                   NaN  
4                                                   NaN  
...                                                 ...  
3062  *need to know specific office to link them.. M...  
3063  *need to know specific office to link them.. M...  
3064                                                NaN  
3065                                                NaN  
3066                                                NaN  

[3067 rows x 6 columns]

In [20]:
df2.shape[0]

3067

This dataset has 3067 rows. Let's try to account for each one based on their column values and recommendations for what we should do with them.

In [21]:
df2[df2['Original Referral Text'].isnull()].shape[0]

183

There are 183 rows which have a null value for the "Original Referral Text" column. I would recommend we keep these accounts as they are and try to find out their referral sources.

In [22]:
df2.loc[~(df2['Original Referral Text'].isnull()) & (df2['Alix\'s comments'] == 'Online?')]

Patient ID       Patient Name Referral Source ID  \
201   001RP000006IvWEYA0    Thanh Trang Luu                NaN   
202   001RP000006IvWFYA0             Vy Luu                NaN   
203   001RP000006IvWGYA0         Trinh Tran                NaN   
293   001RP000006IwlgYAC       Kelsey Allen                NaN   
308   001RP000006IwlvYAC       Bianca Perez                NaN   
...                  ...                ...                ...   
2712  001RP000006JCpiYAG     ALONDRA GARCIA                NaN   
2765  001RP000006JCqbYAG  Ruth Frederickson                NaN   
2807  001RP000006JCrHYAW       JEFFREY WOOD                NaN   
2994  001RP000006JCvtYAG         JACEY LANE                NaN   
3018  001RP000006JCwHYAW      JOHN HUMPHREY                NaN   

        Original Referral Text Preferred Referral Source Alix's comments  
201                   Sonny Do                       NaN         Online?  
202                   Sonny Do                       NaN         Online?  
203                   Sonny Do                       NaN         Online?  
293                       Mike                       NaN         Online?  
308            Marketing Team?                       NaN         Online?  
...                        ...                       ...             ...  
2712            Luis Martinez:                       NaN         Online?  
2765                   Call In                       NaN         Online?  
2807                    Lauren                       NaN         Online?  
2994                    Lauren                       NaN         Online?  
3018  Attorney/ Lily Marketing                       NaN         Online?  

[62 rows x 6 columns]

These are accounts which do have an original referral text, no preferred referral source name, and the comment "Online?" made by Alix. We should follow up with Alix and see if we can classify these accounts as "Online".

In [23]:
df2.loc[~(df2['Original Referral Text'].isnull()) & (df2['Alix\'s comments'] == 'Online?')].shape[0]

62

There are 62 such accounts.

In [24]:
df2.loc[~(df2['Preferred Referral Source'].isnull()) & (df2['Alix\'s comments'] != 'Online?') & (df2['Referral Source ID'].isnull())]

Patient ID               Patient Name Referral Source ID  \
183   001RP000006IvVwYAK                    DOAN DO                NaN   
184   001RP000006IvVxYAK                 Chong Bang                NaN   
185   001RP000006IvVyYAK            Monica Al-Masri                NaN   
187   001RP000006IvW0YAK             Payton Simpson                NaN   
189   001RP000006IvW2YAK             Ashlee Profitt                NaN   
...                  ...                        ...                ...   
3061  001RP000006JCwzYAG   Angel Mota Lopez (minor)                NaN   
3062  001RP000006JCx0YAG  Johana Mota Lopez (minor)                NaN   
3063  001RP000006JCx1YAG   Eliza Mota Lopez (minor)                NaN   
3064  001RP000006JCx2YAG               Rosa Sanchez                NaN   
3065  001RP000006JCx3YAG    Nilofar Nikpour-Farrokh                NaN   

                                 Original Referral Text  \
183                                       Prior Patient   
184                                   Returning Patient   
185                                       Online Search   
187                                              Online   
189                             Marketing Team Espinoza   
...                                                 ...   
3061  Yesenia C. Seguro Noreste, Angel Was Previous ...   
3062                          Yesenia C. Seguro Noreste   
3063                          Yesenia C. Seguro Noreste   
3064                                   Previous Patient   
3065                                   Previous Patient   

      Preferred Referral Source  \
183              Return Patient   
184              Return Patient   
185                      Online   
187                      Online   
189   AC Marketing Team: Miguel   
...                         ...   
3061            Seguro Noroeste   
3062            Seguro Noroeste   
3063            Seguro Noroeste   
3064             Return Patient   
3065             Return Patient   

                                        Alix's comments  
183                                                 NaN  
184                                                 NaN  
185                                                 NaN  
187                                                 NaN  
189                                                 NaN  
...                                                 ...  
3061  *need to know specific office to link them.. M...  
3062  *need to know specific office to link them.. M...  
3063  *need to know specific office to link them.. M...  
3064                                                NaN  
3065                                                NaN  

[1919 rows x 6 columns]

These are the accounts which have preferred referral source with no IDs. Let's come up with a plan to account for these accounts.

In [25]:
df2.loc[~(df2['Preferred Referral Source'].isnull()) & (df2['Alix\'s comments'] != 'Online?') & (df2['Referral Source ID'].isnull())]['Preferred Referral Source'].unique()

array(['Return Patient', 'Online', 'AC Marketing Team: Miguel',
       'Attorney', 'AC Marketing Team: Andrea', 'Patient Referral',
       'Google Appointment', 'Gatti Law Firm',
       'AC Marketing Team: Estefania', 'Alpine Law Group', 'Insurance',
       'The Advocates Injury Attorneys', 'Primary Care Provider',
       'ACS Employee Referral', 'Drive By/Walk In', 'Avrek Law Firm',
       'Transfer', 'AFC Urgent Care', 'LeDoux Insurance',
       'AC Marketing Team: Lily', 'The Joint Chiropractic', 'Yelp',
       'ACS Employee', 'AC Marketing Team: Yesenia',
       'AC Marketing Team: Karla', 'AC Marketing Team: Alma', 'Other',
       'Vietnamese Newspaper', 'AC Marketing Team: Nancy',
       'Seguro Noroeste', 'ZoomCare', 'Columbia Urgent Care',
       'Medical Office', 'AC Marketing Team: Alix', 'Youd Law',
       'AC Marketing Team: Dr. Alexis', 'Concentra',
       'AC Marketing Team: Dr. Linda', 'Auto Body',
       'BestMed (Clinic Unknown)', 'Bishop Legal', 'Bond Legal',
       '

These are the referral sources which do not have referral source IDs.

In [26]:
len(df2.loc[~(df2['Preferred Referral Source'].isnull()) & (df2['Alix\'s comments'] != 'Online?') & (df2['Referral Source ID'].isnull())]['Preferred Referral Source'].unique())

57

There are 57 such referral sources.

In [27]:
nmn_updated = pd.read_excel('C:\\Users\\19713\\Desktop\\ACS\\ACS Account Fix\\Need Marketer Names.xlsx', index_col = 0)

In [28]:
nmn_updated

Referral Sources                  Marketer Name  \
0               Return Patient      Unknown/Other (See notes)   
1                       Online      Unknown/Other (See notes)   
2    AC Marketing Team: Miguel                Miguel Espinoza   
3                     Attorney      Unknown/Other (See notes)   
4    AC Marketing Team: Andrea                      Andrea E.   
..                         ...                            ...   
56                Summit Spine  Regional Director (See notes)   
57  East Canyon Auto Collision                     Alexis Lee   
58   Man M Vu, Attorney at Law                            NaN   
59            Brumley Law Firm  Regional Director (See notes)   
60            Levin and Levine  Regional Director (See notes)   

                                                Notes  
0   No marketer will be attached to this account. ...  
1   No marketer will be attached to this account. ...  
2                                                 NaN  
3   I would like to be notified when this option i...  
4                                                 NaN  
..                                                ...  
56  No marketer will be attached to this account. ...  
57  This is an odd account. I think Alexis knows t...  
58  Already has an account in SalesForce under Man...  
59  No marketer will be attached to this account. ...  
60  No marketer will be attached to this account. ...  

[61 rows x 3 columns]

The dataset in the cell above contain referral sources that do not have referral source IDs, along with the marketer name (if available), and notes provided by Alix.

In [29]:
create_accounts = [9, 11, 15, 27, 35, 36, 41, 42, 43, 48, 49, 52, 53, 56, 57, 59, 60]

In [30]:
nmn_updated.iloc[create_accounts, 1].unique()

array(['Regional Director (See notes)', 'Alexis Lee (See note)',
       'Alexis Lee'], dtype=object)

In [31]:
nmn_updated.iloc[create_accounts, 2].unique()

array(['No marketer will be attached to this account. I believe the regional directors have met with/know of this law firm, so it might be good to create another marketer for "Regional Director" so we can input their notes about certain attorneys',
       'This is kind of an odd source. Either Teresa, Alexis, or Linda pay for the ads in the newspaper. I think it would be best to just create an account for Alexis and put it under her',
       '*Need to create an account for her',
       'No marketer will be attached to this account. This is one of the specialists we refer to, we can list is as Regional Director because they are the ones that maintain the relationship with the specialist.',
       "This is an odd account. I think Alexis knows the owner... Usually patients are referred to Beaverton via text to Thu Lan (I'm not sure if it is the body shop or Alexis texting Thu Lan about the patient)"],
      dtype=object)

Alix has suggested that we create accounts called "Alexis Lee" and "Regional Directors".

In [32]:
len(df2[df2['Preferred Referral Source'].isin(list(nmn_updated.iloc[create_accounts, 0].values))])

50

Once we create accounts for "Alexis Lee" and "Regional Directors", we will be able to provide referral source IDs for 50 accounts in our dataset "df2".

In [33]:
s = set(list(nmn_updated.index))
non_create_accounts = [x for x in s if x not in create_accounts]
nmn_updated_2 = pd.DataFrame(nmn_updated.iloc[non_create_accounts, :])
nmn_updated_2.reset_index(inplace = True, drop = True)

In [34]:
nmn_updated_2[nmn_updated_2['Notes'].isnull()]

Referral Sources      Marketer Name Notes
2        AC Marketing Team: Miguel    Miguel Espinoza   NaN
4        AC Marketing Team: Andrea          Andrea E.   NaN
8     AC Marketing Team: Estefania  Estefania Nateras   NaN
16         AC Marketing Team: Lily           Lily Pak   NaN
20      AC Marketing Team: Yesenia    Yesenia Cardona   NaN
21        AC Marketing Team: Karla        Karla Reyes   NaN
22         AC Marketing Team: Alma        Alma Quiroz   NaN
25        AC Marketing Team: Nancy     Nancy Guardado   NaN
30         AC Marketing Team: Alix      Alix Portlock   NaN
32    AC Marketing Team: Dr. Linda       Linda Nguyen   NaN
37  Saint Moscati Community Health        Alma Quiroz   NaN

Since these are marketer names provided with no notes written for these referral sources, we can go ahead and create these referral sources.

In [35]:
len(df2[df2['Preferred Referral Source'].isin(list(nmn_updated_2[nmn_updated_2['Notes'].isnull()]['Referral Sources'].unique()))])

533

Once we create these referral sources and extract their referral source IDs, we can map them into our dataset "df2", which will fill in 533 entries in the "Referral Source ID" column.

In [36]:
df4 = pd.DataFrame(nmn_updated_2.loc[(~nmn_updated_2['Notes'].isnull()) & ~(nmn_updated_2['Marketer Name'].isnull())].sort_values(by = 'Marketer Name'))

In [37]:
df4.reset_index(inplace = True, drop = True)

In [38]:
df4.to_excel('df4.xlsx')

In [39]:
df4

Referral Sources              Marketer Name  \
0                        ZoomCare                 Alexis Lee   
1            Columbia Urgent Care              Alix Portlock   
2                 AFC Urgent Care              Alix Portlock   
3         Lancaster Family Health          Estefania Nateras   
4                LeDoux Insurance          Estefania Nateras   
5                  Return Patient  Unknown/Other (See notes)   
6   Salem Health (Clinic Unknown)  Unknown/Other (See notes)   
7                      Apple Maps  Unknown/Other (See notes)   
8        BestMed (Clinic Unknown)  Unknown/Other (See notes)   
9                       Auto Body  Unknown/Other (See notes)   
10                      Concentra  Unknown/Other (See notes)   
11                 Medical Office  Unknown/Other (See notes)   
12                   ACS Employee  Unknown/Other (See notes)   
13                          Other  Unknown/Other (See notes)   
14                   ACS Employee  Unknown/Other (See notes)   
15                           Yelp  Unknown/Other (See notes)   
16         The Joint Chiropractic  Unknown/Other (See notes)   
17                       Transfer  Unknown/Other (See notes)   
18               Drive By/Walk In  Unknown/Other (See notes)   
19          ACS Employee Referral  Unknown/Other (See notes)   
20          Primary Care Provider  Unknown/Other (See notes)   
21                      Insurance  Unknown/Other (See notes)   
22                 Gatti Law Firm  Unknown/Other (See notes)   
23             Google Appointment  Unknown/Other (See notes)   
24               Patient Referral  Unknown/Other (See notes)   
25                       Attorney  Unknown/Other (See notes)   
26                         Online  Unknown/Other (See notes)   
27        Legacy (Clinic Unknown)  Unknown/Other (See notes)   
28                       Facebook  Unknown/Other (See notes)   
29                Seguro Noroeste            Yesenia Cardona   

                                                Notes  
0   Can we change this to "ZoomCare (Location Unkn...  
1   Can we change this to "Columbia Urgent Care (L...  
2   Can we change this to "AFC Urgent Care (Locati...  
3   This will be a "blanket" account for the vario...  
4   This will be a "blanket" account for the vario...  
5   No marketer will be attached to this account. ...  
6   I would like to be notified when this option i...  
7   No marketer will be attached to this account. ...  
8   I would like to be notified when this option i...  
9   I would like to be notified when this option i...  
10  I would like to be notified when this option i...  
11  I would like to be notified when this option i...  
12  No marketer will be attached to this account. ...  
13  I would like to be notified when this option i...  
14  No marketer will be attached to this account. ...  
15  No marketer will be attached to this account. ...  
16  Can we change this to "The Joint Chiropractic ...  
17  No marketer will be attached to this account. ...  
18  No marketer will be attached to this account. ...  
19  I would like to be notified when this option i...  
20  I would like to be notified when this option i...  
21  I would like to be notified when this option i...  
22  I would like to be notified when this option i...  
23  No marketer will be attached to this account. ...  
24  No marketer will be attached to this account. ...  
25  I would like to be notified when this option i...  
26  No marketer will be attached to this account. ...  
27  I would like to be notified when this option i...  
28  I would like to be notified when this option i...  
29  This will be a "blanket" account for the vario...

In [40]:
df4[df4['Marketer Name'] != 'Unknown/Other (See notes)']

Referral Sources      Marketer Name  \
0                  ZoomCare         Alexis Lee   
1      Columbia Urgent Care      Alix Portlock   
2           AFC Urgent Care      Alix Portlock   
3   Lancaster Family Health  Estefania Nateras   
4          LeDoux Insurance  Estefania Nateras   
29          Seguro Noroeste    Yesenia Cardona   

                                                Notes  
0   Can we change this to "ZoomCare (Location Unkn...  
1   Can we change this to "Columbia Urgent Care (L...  
2   Can we change this to "AFC Urgent Care (Locati...  
3   This will be a "blanket" account for the vario...  
4   This will be a "blanket" account for the vario...  
29  This will be a "blanket" account for the vario...

In [41]:
df4[df4['Marketer Name'] != 'Unknown/Other (See notes)']['Notes'].unique()

array(['Can we change this to "ZoomCare (Location Unknown)"?This will be a "blanket" account for the various offices because a lot of the time the patient does not specify which location referred them ',
       'Can we change this to "Columbia Urgent Care (Location Unknown)"?This will be a "blanket" account for the various offices because a lot of the time the patient does not specify which location referred them ',
       'Can we change this to "AFC Urgent Care (Location Unknown)"? This will be a "blanket" account for the various AFC clinics because a lot of the time the patient does not specify which clinic they came from. ',
       'This will be a "blanket" account for the various offices because a lot of the time the patient does not specify which location referred them '],
      dtype=object)

In [42]:
list(df4[df4['Marketer Name'] != 'Unknown/Other (See notes)']['Referral Sources'].values)

['ZoomCare',
 'Columbia Urgent Care',
 'AFC Urgent Care',
 'Lancaster Family Health',
 'LeDoux Insurance',
 'Seguro Noroeste']

Alix has suggested that we make "blanket" accounts for these referral sources.

In [43]:
len(df2[df2['Preferred Referral Source'].isin(list(df4[df4['Marketer Name'] != 'Unknown/Other (See notes)']['Referral Sources'].values))])

54

After creating referral source IDs for the referral sources in the list above, we can account for 54 more entries in the "Referral Source ID" column.

In [44]:
df4[df4['Marketer Name'] == 'Unknown/Other (See notes)']['Notes'].unique()

array(["No marketer will be attached to this account. We can either use the Unknown or Other marketer account. Use whatever account Minh would prefer to use for accounts that won't have a marketer attached",
       'I would like to be notified when this option is selected so I can gather more info from the clinic. This will be a "blanket" account for the various offices because a lot of the time the patient does not specify which location referred them ',
       'I would like to be notified when this option is selected so I can gather more info from the clinic. This account will be a "blanket" account, so we can either use the Unknown or Other marketer account. We can use whatever account Minh would prefer to use for accounts that won\'t have a marketer attached',
       'Can we change this to "The Joint Chiropractic (Location Unknown)"? I would like to be notified when this option is selected so I can gather more info from the clinic. This account will be a "blanket" account, so we ca

In [45]:
df4[df4['Notes'] == "No marketer will be attached to this account. We can either use the Unknown or Other marketer account. Use whatever account Minh would prefer to use for accounts that won't have a marketer attached"]

Referral Sources              Marketer Name  \
5       Return Patient  Unknown/Other (See notes)   
7           Apple Maps  Unknown/Other (See notes)   
12        ACS Employee  Unknown/Other (See notes)   
14        ACS Employee  Unknown/Other (See notes)   
15                Yelp  Unknown/Other (See notes)   
17            Transfer  Unknown/Other (See notes)   
18    Drive By/Walk In  Unknown/Other (See notes)   
23  Google Appointment  Unknown/Other (See notes)   
24    Patient Referral  Unknown/Other (See notes)   
26              Online  Unknown/Other (See notes)   

                                                Notes  
5   No marketer will be attached to this account. ...  
7   No marketer will be attached to this account. ...  
12  No marketer will be attached to this account. ...  
14  No marketer will be attached to this account. ...  
15  No marketer will be attached to this account. ...  
17  No marketer will be attached to this account. ...  
18  No marketer will be attached to this account. ...  
23  No marketer will be attached to this account. ...  
24  No marketer will be attached to this account. ...  
26  No marketer will be attached to this account. ...

For these referral sources, we can use either the "Unknown" or "Other" marketer account.

In [46]:
list(df4[df4['Notes'] == "No marketer will be attached to this account. We can either use the Unknown or Other marketer account. Use whatever account Minh would prefer to use for accounts that won't have a marketer attached"]['Referral Sources'].unique())

['Return Patient',
 'Apple Maps',
 'ACS Employee',
 'Yelp',
 'Transfer',
 'Drive By/Walk In',
 'Google Appointment',
 'Patient Referral',
 'Online']

In [47]:
df2[df2['Preferred Referral Source'].isin(list(df4[df4['Notes'] == "No marketer will be attached to this account. We can either use the Unknown or Other marketer account. Use whatever account Minh would prefer to use for accounts that won't have a marketer attached"]['Referral Sources'].unique()))].shape[0]

1070

Creating referral source IDs for these referral sources would account for 1070 accounts.

In [48]:
other_notes = ['I would like to be notified when this option is selected so I can gather more info from the clinic. This will be a "blanket" account for the various offices because a lot of the time the patient does not specify which location referred them ',
       'I would like to be notified when this option is selected so I can gather more info from the clinic. This account will be a "blanket" account, so we can either use the Unknown or Other marketer account. We can use whatever account Minh would prefer to use for accounts that won\'t have a marketer attached',
       'Can we change this to "The Joint Chiropractic (Location Unknown)"? I would like to be notified when this option is selected so I can gather more info from the clinic. This account will be a "blanket" account, so we can either use the Unknown or Other marketer account. We can use whatever account Minh would prefer to use for accounts that won\'t have a marketer attached',
       'I would like to be notified when this option is selected so I can gather the attorney\'s name from the clinic. There are a number of Gatti law offices that various marketers visit, so this account will be a "blanket" account. We can either use the Unknown or Other marketer account for this account. Use whatever account Minh would prefer to use for accounts that won\'t have a marketer attached',
       'I would like to be notified when this option is selected so I can gather the Attorney\'s name from the clinic. Since this account will be a "blanket" account for the various Attorneys that refer that no marketer is attached to, we can either use the Unknown or Other marketer account. We can use whatever account Minh would prefer to use for accounts that won\'t have a marketer attached']

In [49]:
list(df4[df4['Notes'].isin(other_notes)]['Referral Sources'].unique())

['Salem Health (Clinic Unknown)',
 'BestMed (Clinic Unknown)',
 'Auto Body',
 'Concentra',
 'Medical Office',
 'Other',
 'The Joint Chiropractic',
 'ACS Employee Referral',
 'Primary Care Provider',
 'Insurance',
 'Gatti Law Firm',
 'Attorney',
 'Legacy (Clinic Unknown)',
 'Facebook']

In [50]:
df2[df2['Preferred Referral Source'].isin(list(df4[df4['Notes'].isin(other_notes)]['Referral Sources'].unique()))]['Preferred Referral Source'].unique()

array(['Attorney', 'Gatti Law Firm', 'Insurance', 'Primary Care Provider',
       'ACS Employee Referral', 'The Joint Chiropractic', 'Other',
       'Medical Office', 'Concentra', 'Auto Body',
       'BestMed (Clinic Unknown)', 'Salem Health (Clinic Unknown)',
       'Legacy (Clinic Unknown)', 'Facebook'], dtype=object)

These are the referral sources for which when they are selected, Alix would like to receive a notification so that she can follow up to find more information. For the time being we can use the "Unknown" or "Other" marketer option in order to create these referral sources.

In [51]:
df2[df2['Preferred Referral Source'].isin(list(df4[df4['Notes'].isin(other_notes)]['Referral Sources'].unique()))]['Preferred Referral Source'].shape[0]

212

Creating ID numbers for these referral sources would account for 212 accounts.

In [52]:
50 + 533 + 54 + 1070 + 212 == df2.loc[(~df2['Preferred Referral Source'].isnull()) & (df2['Referral Source ID'].isnull())].shape[0]

True

We have successfully accounted for all patient accounts which did not have a referral source ID.

# Conclusion

Here are my recommendations for this project:

1. Find referral sources for the following patients:

In [55]:
df2[df2['Original Referral Text'].isnull()]['Patient Name'].unique()

array(['Freddie Elliott JR', 'test tester', 'Derrisha Sutton',
       '(Yuri) Yuriel Leyva', 'Abdullahi Ali', 'Adriana Flores-Arteaga',
       'Adriana Magallow Vaga', 'Alexandra Parker', 'Andy Torres-Garcia',
       'Angel Le', 'Anna Henning', 'Antonio Rodriguez',
       'Arturo Rodriguez', 'Belkys Caballero', 'Biyod Sheikhahmell',
       'Blanca Castellanos', 'Bogdan Shaban', 'Chantel Siduemoto',
       'Charles Freeborn', 'Chelsey Saeteurn', 'Christopher Hess',
       'Cristian Villareal', 'Crystal Johnson', 'Daniel Jones',
       'Dao Ittihrit', 'Delia Avila Chavez', 'Desmond Washington',
       'Desmond Washington JR', 'Diem Truong', 'Dominique Stravghter',
       'Donvan Thai', 'Duc Huu Nguyen', 'Edgar Pantoja', 'Edgar Villegas',
       'Emily Parker', 'Ethan Tran', 'Evelyn Kelly', 'Faisal Khurshid',
       'Fernando Albarran Dinh', 'Gabriel Fish', 'Gerry Fridlund',
       'Giang Bui', 'Hana Mulah', 'Hanah Nguyen', 'Hang Thai',
       'Hatem Boukhari', 'Hector Calderon', 'Herbert

2. Verify that the following patients were verified online.

In [57]:
df2[df2['Preferred Referral Source'].isnull() & ~(df2['Original Referral Text'].isnull())]['Original Referral Text'].unique()

array(['Sonny Do', 'Mike', 'Marketing Team?', 'Work Relation',
       "Dr. Matt'S Hop Festival Committee Board", 'Employee - Nectar',
       'Marketing Team: Sarah', 'Call In', 'Job', 'Staff',
       'Wrote Internet But May Have Been Referred By Priority Life Chiro',
       'Rose Urgent Care: Sarah', 'Accident Care Se Front Desk', 'Ad',
       'Chamber Of Commerce', 'Sara: Marketing Team', 'Referral',
       'Beaverton Office Gave Me Info', 'Attorney (Tony)', 'Work', 'Tony',
       'Wife/Google/State Farm', 'R&R Velma Dr. Matt Referral Group',
       'Referral By Tony', 'Attorney: David Wales', 'Lazer Augi (Boss)',
       'Attorney- Sonny', 'Lauren Neuman', 'Jason Hargrave',
       'Aleida Gonzalez / Autobody Shop (Villegas Autobody)',
       'Rodrigo Morales', 'Lorena Santana', 'Kelly Whitlok',
       'Luis Martinez:', 'Lauren', 'Attorney/ Lily Marketing'],
      dtype=object)

3. Create referral sources called "Alexis Lee", "Regional Directors", and sources for the following list:

In [59]:
nmn_updated_2[nmn_updated_2['Notes'].isnull()]['Referral Sources'].unique()

array(['AC Marketing Team: Miguel', 'AC Marketing Team: Andrea',
       'AC Marketing Team: Estefania', 'AC Marketing Team: Lily',
       'AC Marketing Team: Yesenia', 'AC Marketing Team: Karla',
       'AC Marketing Team: Alma', 'AC Marketing Team: Nancy',
       'AC Marketing Team: Alix', 'AC Marketing Team: Dr. Linda',
       'Saint Moscati Community Health'], dtype=object)

4. Create sources for these accounts using the "Unknown" or "Other" marketer option.

In [62]:
df4[df4['Notes'] == "No marketer will be attached to this account. We can either use the Unknown or Other marketer account. Use whatever account Minh would prefer to use for accounts that won't have a marketer attached"]['Referral Sources'].unique()

array(['Return Patient', 'Apple Maps', 'ACS Employee', 'Yelp', 'Transfer',
       'Drive By/Walk In', 'Google Appointment', 'Patient Referral',
       'Online'], dtype=object)

5. Create blanket accounts for the following sources:

In [60]:
list(df4[df4['Marketer Name'] != 'Unknown/Other (See notes)']['Referral Sources'].values)

['ZoomCare',
 'Columbia Urgent Care',
 'AFC Urgent Care',
 'Lancaster Family Health',
 'LeDoux Insurance',
 'Seguro Noroeste']

6. Create referral sources for the following accounts with the condition that Alix receive a notification when selected so that she can follow up and gather more information.

In [63]:
df2[df2['Preferred Referral Source'].isin(list(df4[df4['Notes'].isin(other_notes)]['Referral Sources'].unique()))]['Preferred Referral Source'].unique()

array(['Attorney', 'Gatti Law Firm', 'Insurance', 'Primary Care Provider',
       'ACS Employee Referral', 'The Joint Chiropractic', 'Other',
       'Medical Office', 'Concentra', 'Auto Body',
       'BestMed (Clinic Unknown)', 'Salem Health (Clinic Unknown)',
       'Legacy (Clinic Unknown)', 'Facebook'], dtype=object)